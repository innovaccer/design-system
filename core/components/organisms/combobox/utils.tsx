import * as React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  inputTriggerRef?: any,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>,
  multiSelect?: boolean,
  listRef?: any
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption, listRef);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption, listRef);
      break;
    case 'Enter':
      handleEnterKey(focusedOption, multiSelect, inputTriggerRef, listRef, setFocusedOption);
      setHighlightLastItem?.(false);
      setHighlightFirstItem?.(false);
      break;
    case 'Escape':
      setOpenPopover?.(false);
      inputTriggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    default:
      break;
  }
};

const handleEnterKey = (
  focusedOption: Element | undefined,
  multiSelect?: boolean,
  inputTriggerRef?: any,
  listRef?: any,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>
) => {
  (focusedOption as HTMLElement)?.click();
  if (!multiSelect) {
    inputTriggerRef.current.focus();
  } else {
    // to focus first option by default when last option is selected
    const listItems = listRef.current.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
    const listArr = Array.from(listItems);
    const index = listArr.findIndex((item) => {
      return item == focusedOption;
    });

    if (index === listArr.length - 1) {
      (listItems[0] as HTMLElement).focus();
      setFocusedOption && setFocusedOption(listItems[0]);
      listItems[0].scrollIntoView({ block: 'center' });
    }
  }
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any
) => {
  const listItems = listRef.current.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  if (index === -1) {
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else {
    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;
  }

  const targetOption = listItems[index];

  (targetOption as HTMLElement).focus();
  setFocusedOption && setFocusedOption(targetOption);
  targetOption?.scrollIntoView?.({ block: 'center' });
};
