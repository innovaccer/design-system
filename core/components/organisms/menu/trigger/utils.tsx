import * as React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean | undefined>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      setOpenPopover?.(true);
      setHighlightLastItem?.(true);
      break;
    case 'ArrowDown':
      event.preventDefault();
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'Escape':
    case 'Tab':
      setOpenPopover?.(false);
      break;
    default:
      break;
  }
};

export const focusListItem = (
  position: string,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any
) => {
  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  if (position === 'down') {
    targetOption = listItems?.[0];
  } else {
    targetOption = listItems?.[listItems.length - 1];
  }
  (targetOption as HTMLElement)?.focus();

  if (targetOption && typeof targetOption.scrollIntoView === 'function') {
    (targetOption as HTMLElement)?.scrollIntoView({ block: 'end' });
  }
  setFocusedOption && setFocusedOption(targetOption);
};
