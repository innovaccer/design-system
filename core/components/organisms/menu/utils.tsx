import React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean | undefined>>,
  menuTriggerRef?: React.RefObject<HTMLButtonElement | null>,
  listRef?: React.RefObject<HTMLDivElement | null>,
  subListRef?: React.RefObject<HTMLDivElement | null> | null,
  isSubMenuTrigger?: boolean,
  triggerRef?: React.RefObject<HTMLDivElement | null> | React.MutableRefObject<HTMLDivElement | null>,
  menuID?: string,
  triggerID?: string,
  parentListRef?: React.RefObject<HTMLDivElement | null> | null
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption, listRef);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption, listRef);
      break;
    case 'Enter':
      (focusedOption as HTMLElement)?.click();
      setOpenPopover?.(false);
      break;
    case 'Escape':
      setOpenPopover?.(false);
      if (triggerRef && !isSubMenuTrigger) {
        triggerRef?.current?.focus();
      } else {
        menuTriggerRef?.current?.focus();
      }
      setFocusedOption?.(undefined);
      break;
    case 'Tab':
      setOpenPopover?.(false);
      break;
    case 'ArrowRight':
      navigateSubMenu(isSubMenuTrigger, 'right', subListRef, menuID, triggerID, parentListRef);
      break;
    case 'ArrowLeft':
      navigateSubMenu(isSubMenuTrigger, 'left', subListRef, menuID, triggerID, parentListRef);
      break;
    default:
      break;
  }
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any
) => {
  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  if (index === -1) {
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else {
    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;
  }

  const targetOption = listItems[index];
  (targetOption as HTMLElement).focus();
  setFocusedOption && setFocusedOption(targetOption);
  targetOption?.scrollIntoView?.({ block: 'center' });
};

const navigateSubMenu = (
  isSubMenuTrigger?: boolean,
  direction?: string,
  subListRef?: React.RefObject<HTMLDivElement | null> | null,
  menuID?: string,
  triggerID?: string,
  parentListRef?: React.RefObject<HTMLDivElement | null> | null
) => {
  const element = document.querySelector(`[data-name="${menuID}"]`);
  const menuPlacement = element?.getAttribute('data-placement');

  if (isSubMenuTrigger) {
    if (
      (direction === 'right' && menuPlacement?.includes('right')) ||
      (direction === 'left' && menuPlacement?.includes('left'))
    ) {
      const listItems = subListRef?.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
      (listItems?.[0] as HTMLElement).focus();
    }
  } else if (
    (direction === 'left' && menuPlacement?.includes('right')) ||
    (direction === 'right' && menuPlacement?.includes('left'))
  ) {
    const triggerElement = parentListRef?.current?.querySelector(`#${triggerID}`)?.firstChild;
    (triggerElement as HTMLElement)?.focus();
  }
};
