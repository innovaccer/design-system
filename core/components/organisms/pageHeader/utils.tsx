import * as React from 'react';
import classNames from 'classnames';
import { navigationPositionType } from './PageHeader';
import { Heading, Column } from '@/index';
import styles from '@css/components/pageHeader.module.css';

export const Status = (props: {
  status: React.ReactNode;
  meta: React.ReactNode;
  navigationPosition: navigationPositionType;
  navigation: React.ReactNode;
  tabs: React.ReactNode;
}) => {
  const { status, meta, navigationPosition, navigation, tabs } = props;

  const statusClasses = classNames({
    [styles['PageHeader-statusWrapper']]: true,
    'mb-3': (navigationPosition === 'bottom' && navigation) || tabs,
  });

  return (
    <>
      {(status || meta) && (
        <div className={statusClasses} data-test="DesignSystem-PageHeader--Status">
          {status}
          {meta}
        </div>
      )}
    </>
  );
};

export const Action = (props: { actions: React.ReactNode; navigation: React.ReactNode; stepper: React.ReactNode }) => {
  const { actions, navigation, stepper } = props;

  return (
    <>
      {actions ? (
        <Column size="4" sizeXL="4" sizeM="4" data-test="DesignSystem-PageHeader--Actions">
          <div className={styles['PageHeader-actionsWrapper']}>{actions}</div>
        </Column>
      ) : (
        (navigation || stepper) && (
          <Column size="4" sizeXL="4" sizeM="4" data-test="DesignSystem-PageHeader--Actions">
            <div className={styles['PageHeader-actionsWrapper']}></div>
          </Column>
        )
      )}
    </>
  );
};

export const Nav = (props: { navigation: React.ReactNode; stepper: React.ReactNode }) => {
  const { navigation, stepper } = props;

  if (!navigation && !stepper) {
    return null;
  }
  return (
    <div className={styles['PageHeader-navigationWrapper']} data-test="DesignSystem-PageHeader--Nav">
      {navigation || stepper}
    </div>
  );
};

export const CenterNav = (props: {
  colSize: string;
  breadcrumbs: React.ReactNode;
  navigationPosition: navigationPositionType;
  navigation: React.ReactNode;
  stepper: React.ReactNode;
}) => {
  const { colSize, breadcrumbs, navigationPosition } = props;
  return (
    <>
      {(!breadcrumbs || navigationPosition === 'center') && colSize === '4' && (
        <Column size="4" sizeXL="4" sizeM="4" data-test="DesignSystem-PageHeader--CenterNav">
          <Nav {...props} />
        </Column>
      )}
    </>
  );
};

export const BackButton = (props: { button: React.ReactNode }) => {
  const { button } = props;
  return (
    <>
      {button && (
        <div className="mr-5 my-3" data-test="DesignSystem-PageHeader--Button">
          {button}
        </div>
      )}
    </>
  );
};

export const Title = (props: { badge: React.ReactNode; title: string }) => {
  const { badge, title } = props;
  return (
    <div className={styles['PageHeader-titleWrapper']} data-test="DesignSystem-PageHeader--Title">
      <Heading className={styles['PageHeader-title']}>{title}</Heading>
      {badge}
    </div>
  );
};
