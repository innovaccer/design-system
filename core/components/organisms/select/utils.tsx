import * as React from 'react';
import { OptionType } from '@/common.type';

export const mapInitialValue = (multiSelect: boolean, selectedValue: OptionType | OptionType[] | undefined) => {
  if (multiSelect) {
    return selectedValue && !Array.isArray(selectedValue) ? [selectedValue] : selectedValue || [];
  } else {
    return selectedValue || { label: '', value: '' };
  }
};

export const elementExist = (targetObject: OptionType, mainList: OptionType | OptionType[] | undefined) => {
  if (!Array.isArray(mainList)) {
    return targetObject.value === mainList?.value ? 0 : -1;
  }
  return mainList.findIndex((item) => item.value === targetObject.value);
};

export const removeOrAddToList = (targetObject: OptionType, prevList: OptionType[]) => {
  const newList = [...prevList];
  const existingIndex = elementExist(targetObject, newList);
  if (existingIndex !== -1) {
    newList.splice(existingIndex, 1);
  } else {
    newList.push(targetObject);
  }
  return newList;
};

export const computeValue = (
  multiSelect: boolean | undefined,
  selectValue: any,
  setLabel?: (count: number) => string | undefined
) => {
  if (!multiSelect) {
    return selectValue?.label?.trim() || '';
  }
  const label = setLabel?.(selectValue.length);

  if (label) {
    return label;
  }

  if (selectValue.length <= 2) {
    return selectValue.map((pair: any) => `${pair.label}`).join(', ');
  } else {
    return `${selectValue.length} Selected`;
  }
};

export const handleKeyDownTrigger = (
  event: React.KeyboardEvent,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>
) => {
  switch (event.key) {
    case 'Enter':
      event.preventDefault();
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowDown':
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowUp':
      setOpenPopover?.(true);
      setHighlightLastItem?.(true);
      break;
    default:
      break;
  }
};

export const focusListItem = (
  position: string,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any
) => {
  const searchInput = listRef.current?.querySelectorAll('[data-test="DesignSystem-Select--Input"]');
  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  if (position === 'down') {
    targetOption = searchInput?.[0] || listItems?.[0];
  } else {
    targetOption = listItems?.[listItems.length - 1];
  }

  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>,
  listRef?: any,
  withSearch?: boolean,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  triggerRef?: any
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'Enter':
      handleEnterKey(focusedOption);
      setHighlightLastItem?.(false);
      setHighlightFirstItem?.(false);
      break;
    case 'Tab':
      setHighlightLastItem?.(false);
      setHighlightFirstItem?.(false);
      break;
    case 'Escape':
      setOpenPopover?.(false);
      triggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    default:
      break;
  }
};

export const handleEnterKey = (focusedOption: Element | undefined) => {
  (focusedOption as HTMLElement)?.click();
};

export const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any,
  withSearch?: boolean
) => {
  const listItems = listRef.current.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  if (index === -1) {
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else if (
    (withSearch && index === 0 && direction === 'up') ||
    (withSearch && index === listItems.length - 1 && direction === 'down')
  ) {
    const searchInput = listRef.current.querySelector('[data-test="DesignSystem-Select--Input"]');
    searchInput.focus();
    setFocusedOption && setFocusedOption(searchInput);
  } else {
    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;

    const targetOption = listItems[index];

    (targetOption as HTMLElement).focus();
    setFocusedOption && setFocusedOption(targetOption);
    targetOption.scrollIntoView({ block: 'center' });
  }
};

export const handleInputKeyDown = (
  event: React.KeyboardEvent,
  listRef: any,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  triggerRef?: any
) => {
  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      targetOption = listItems[listItems.length - 1];
      break;
    case 'ArrowDown':
      event.preventDefault();
      targetOption = listItems[0];
      break;
    case 'Escape':
      setOpenPopover?.(false);
      triggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    default:
      break;
  }

  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
