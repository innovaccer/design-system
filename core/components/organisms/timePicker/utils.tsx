import { TimeType, AMPMType } from './TimePicker';
import { InputProps } from '@/index.type';

interface TimeObject {
  hours: number;
  minutes: number;
  am_pm: string;
}

export const placeholders: { [key: string]: InputProps['placeholder'] } = {
  ['hh:mm']: '--:--',
  ['hh:mm AM']: '--:-- AM',
};

export const isPlaceholderPresent = (placeholderChar: string, time?: string) => {
  return time && time.includes(placeholderChar);
};

export const isFormat12hour = (format: string) => {
  return format === 'hh:mm AM';
};

const get12hourFormat = (hours: number) => {
  const AMPM = hours < 12 ? 'AM' : 'PM';
  const hrs = hours % 12 || 12;
  return { hrs, AMPM };
};

const get24hourFormat = (hours: number, am_pm?: AMPMType | string) => {
  let convertedHours = hours;

  if (am_pm) {
    if (am_pm === 'PM' && hours < 12) {
      convertedHours = hours + 12;
    } else if (am_pm === 'AM' && hours === 12) {
      convertedHours = hours - 12;
    }

    return convertedHours;
  }

  return hours;
};

export const translateToTime = (format: string, time?: TimeType) => {
  if (!time) return '';

  if (typeof time === 'number') {
    const timeObj = getTimeObjectFromNumber(format, time);
    return translateToString(format, timeObj);
  }

  return time;
};

const getTimeObjectFromNumber = (format: string, time: number) => {
  const d = new Date(time);
  const hrs = d.getHours();

  const hours = isFormat12hour(format) ? get12hourFormat(hrs).hrs : hrs;
  const am_pm = isFormat12hour(format) ? get12hourFormat(hrs).AMPM : '';
  const minutes = d.getMinutes();
  const seconds = d.getSeconds();
  return { hours, minutes, seconds, am_pm };
};

export const getTimeObjFromStr = (format: string, time: string) => {
  const separator = ':';
  let hours = 0;
  let minutes = 0;

  const len = format.length;
  const timeLength = time.length;
  const lastChars = format.substring(len - 2, len);
  const is12hrFormat = lastChars === 'AM' || lastChars === 'PM';

  const am_pm: AMPMType | string = is12hrFormat ? time.substring(timeLength - 2, timeLength) : '';
  const timeFormat = is12hrFormat ? time.substring(0, timeLength - 3) : time;
  const inputFormat = is12hrFormat ? format.substring(0, len - 3) : format;

  const v = timeFormat.split(separator);
  inputFormat.split(separator).forEach((f, i) => {
    switch (f) {
      case 'hh':
        hours = +v[i] || 0;
        break;
      case 'mm':
        minutes = +v[i] || 0;
        break;
    }
  });
  return { hours, minutes, am_pm };
};

export const getOutputTimeString = (inputFormat: string, outputFormat: string, time: string) => {
  if (inputFormat === outputFormat) return time;

  const { hours, minutes, am_pm } = getTimeObjFromStr(inputFormat, time);
  const AMPM = isFormat12hour(outputFormat) ? get12hourFormat(hours).AMPM : '';
  const hrs = isFormat12hour(outputFormat) ? get12hourFormat(hours).hrs : get24hourFormat(hours, am_pm);

  const timeStr = translateToString(outputFormat, { minutes, hours: hrs, am_pm: AMPM });
  return timeStr;
};

const translateToString = (format: string, time: TimeObject) => {
  const { hours, minutes, am_pm } = time;

  const separator = ':';
  const timeFormat = format.split(' ');
  const v = timeFormat[0].split(separator);
  let val = '';

  v.forEach((f, i) => {
    switch (f) {
      case 'hh':
        val += hours < 10 ? `0${hours}` : hours;
        break;
      case 'mm':
        val += minutes < 10 ? `0${minutes}` : minutes;
        break;
    }
    if (i !== f.length - 1) val += separator;
  });

  val += isFormat12hour(format) && am_pm ? ` ${am_pm}` : '';
  return val;
};
