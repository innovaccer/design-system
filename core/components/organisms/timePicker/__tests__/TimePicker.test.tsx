import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { testHelper, filterUndefined, valueHelper, testMessageHelper } from '@/utils/testHelper';
import { TimePicker } from '@/index';
import { TimePickerProps as Props } from '@/index.type';

const timeFormat = ['hh:mm AM', 'hh:mm'];
const BooleanValue = [true, false];
const timeValues = [1629880121, '3:30 AM'];
const expectedTimeValue = ['08:44 PM', '3:30 AM'];
const FunctionValue = jest.fn();

describe('TimePicker component prop:inputFormat snapshot', () => {
  const mapper: Record<string, any> = {
    inputFormat: valueHelper(timeFormat, { required: true, iterate: true }),
    onTimeChange: valueHelper(FunctionValue, { required: true })
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(
        <TimePicker
          {...attr}
        />
      );
      expect(baseElement).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('TimePicker component prop:outputFormat snapshot', () => {
  const mapper: Record<string, any> = {
    outputFormat: valueHelper(timeFormat, { required: true, iterate: true }),
    onTimeChange: valueHelper(FunctionValue, { required: true })
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(
        <TimePicker
          {...attr}
        />
      );
      expect(baseElement).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('TimePicker component prop:validators snapshot', () => {
  const mapper: Record<string, any> = {
    validators: valueHelper(BooleanValue, { required: true, iterate: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(
        <TimePicker
          {...attr}
        />
      );
      expect(baseElement).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('TimePicker component prop:time snapshot', () => {
  const mapper: Record<string, any> = {
    time: valueHelper(timeValues, { required: true, iterate: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(
        <TimePicker
          {...attr}
        />
      );
      expect(baseElement).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('TimePicker component with different prop:time values', () => {

  timeValues.forEach((time, key) => {
    it(`check time if prop:time is passed as ${typeof (time)} `, () => {
      const { getByTestId } = render(
        <TimePicker
          time={time}
        />
      );
      expect(getByTestId('DesignSystem-Input')).toHaveValue(expectedTimeValue[key]);
    });
  });
});

describe('TimePicker component with different prop:inputFormat values', () => {
  it('check format if prop:inputFormat is hh:mm', () => {
    const { getByTestId } = render(
      <TimePicker
        time={timeValues[0]}
        inputFormat="hh:mm"
      />
    );
    expect(getByTestId('DesignSystem-Input')).toHaveValue('20:44');
  });

  it('check format if prop:inputFormat is hh:mm AM', () => {
    const { getByTestId } = render(
      <TimePicker
        time={timeValues[0]}
        inputFormat="hh:mm AM"
      />
    );
    expect(getByTestId('DesignSystem-Input')).toHaveValue(expectedTimeValue[0]);
  });
});

describe('TimePicker component event handler', () => {
  it('onBlur Event handler ', () => {
    const { getByTestId } = render(
      <TimePicker
        time={timeValues[1]}
        onTimeChange={FunctionValue}
      />
    );
    const InputTestId = getByTestId('DesignSystem-Input');
    fireEvent.blur(InputTestId);
    expect(FunctionValue).toHaveBeenCalled();
  });

  it('onChange Event handler ', () => {
    const { getByTestId } = render(
      <TimePicker
        time={timeValues[1]}
        onTimeChange={FunctionValue}
      />
    );
    const InputTestId = getByTestId('DesignSystem-Input');
    fireEvent.change(InputTestId, { target: { value: timeValues[0] } });
    expect(FunctionValue).toHaveBeenCalled();
  });

  it('onClear Event handler ', () => {
    const { getByTestId } = render(
      <TimePicker
        time={timeValues[1]}
        onTimeChange={FunctionValue}
      />
    );
    const InputTestId = getByTestId('DesignSystem-Input--closeIcon');
    fireEvent.click(InputTestId);
    expect(FunctionValue).toHaveBeenCalled();
  });
});

describe('TimePicker component placeholder', () => {
  it('check placeholder if inputFormat is default ', () => {
    const { getByTestId } = render(
      <TimePicker
        time={timeValues[1]}
        inputOptions={{
          placeholderChar: '*',
        }}
      />
    );
    const InputTestId = getByTestId('DesignSystem-Input--closeIcon');
    fireEvent.click(InputTestId);
    expect(FunctionValue).toHaveBeenCalled();
    expect(getByTestId('DesignSystem-Input')).toHaveValue('**:** *M');
  });

  it('check placeholder if inputFormat is hh:mm ', () => {
    const { getByTestId } = render(
      <TimePicker
        time={timeValues[1]}
        inputFormat="hh:mm"
        inputOptions={{
          placeholderChar: '*'
        }}
      />
    );
    const InputTestId = getByTestId('DesignSystem-Input--closeIcon');
    fireEvent.click(InputTestId);
    expect(FunctionValue).toHaveBeenCalled();
    expect(getByTestId('DesignSystem-Input')).toHaveValue('**:**');
  });
});

describe('TimePicker component with different time format', () => {
  it('check when inputFormat and outputFormat is different', () => {
    const { getByTestId } = render(
      <TimePicker
        time={timeValues[1]}
        onTimeChange={FunctionValue}
        inputFormat="hh:mm AM"
        outputFormat="hh:mm"
      />
    );
    const InputTestId = getByTestId('DesignSystem-Input');
    fireEvent.blur(InputTestId);
    expect(FunctionValue).toHaveBeenCalled();
  });
});
