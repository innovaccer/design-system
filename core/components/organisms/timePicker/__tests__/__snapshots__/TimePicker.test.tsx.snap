// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker component prop:inputFormat snapshot 
  inputFormat: "hh:mm AM", onTimeChange: "[Function]"
 1`] = `
<body>
  <div>
    <div
      class="d-flex flex-column flex-grow-1"
      data-test="DesignSystem-InputMask--Wrapper"
    >
      <div
        class="Input Input--regular"
        data-test="DesignSystem-InputWrapper"
        role="presentation"
        style="min-width: 256px;"
      >
        <input
          autocomplete="off"
          class="Input-input Input-input--regular"
          data-test="DesignSystem-Input"
          placeholder="--:-- AM"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`TimePicker component prop:inputFormat snapshot 
  inputFormat: "hh:mm", onTimeChange: "[Function]"
 1`] = `
<body>
  <div>
    <div
      class="d-flex flex-column flex-grow-1"
      data-test="DesignSystem-InputMask--Wrapper"
    >
      <div
        class="Input Input--regular"
        data-test="DesignSystem-InputWrapper"
        role="presentation"
        style="min-width: 256px;"
      >
        <input
          autocomplete="off"
          class="Input-input Input-input--regular"
          data-test="DesignSystem-Input"
          placeholder="--:--"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`TimePicker component prop:outputFormat snapshot 
  outputFormat: "hh:mm AM", onTimeChange: "[Function]"
 1`] = `
<body>
  <div>
    <div
      class="d-flex flex-column flex-grow-1"
      data-test="DesignSystem-InputMask--Wrapper"
    >
      <div
        class="Input Input--regular"
        data-test="DesignSystem-InputWrapper"
        role="presentation"
        style="min-width: 256px;"
      >
        <input
          autocomplete="off"
          class="Input-input Input-input--regular"
          data-test="DesignSystem-Input"
          placeholder="--:-- AM"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`TimePicker component prop:outputFormat snapshot 
  outputFormat: "hh:mm", onTimeChange: "[Function]"
 1`] = `
<body>
  <div>
    <div
      class="d-flex flex-column flex-grow-1"
      data-test="DesignSystem-InputMask--Wrapper"
    >
      <div
        class="Input Input--regular"
        data-test="DesignSystem-InputWrapper"
        role="presentation"
        style="min-width: 256px;"
      >
        <input
          autocomplete="off"
          class="Input-input Input-input--regular"
          data-test="DesignSystem-Input"
          placeholder="--:-- AM"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`TimePicker component prop:time snapshot 
  time: "3:30 AM"
 1`] = `
<body>
  <div>
    <div
      class="d-flex flex-column flex-grow-1"
      data-test="DesignSystem-InputMask--Wrapper"
    >
      <div
        class="Input Input--regular"
        data-test="DesignSystem-InputWrapper"
        role="presentation"
        style="min-width: 256px;"
      >
        <input
          autocomplete="off"
          class="Input-input Input-input--regular"
          data-test="DesignSystem-Input"
          placeholder="--:-- AM"
          type="text"
          value="3:30 AM"
        />
        <div
          class="Input-icon Input-iconWrapper--right"
        >
          <i
            class="material-icons material-icons-round Icon Input-icon--right"
            data-test="DesignSystem-Input--closeIcon"
            role="button"
            style="font-size: 16px; width: 16px;"
            tabindex="0"
          >
            close_round
          </i>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TimePicker component prop:time snapshot 
  time: 1629880121
 1`] = `
<body>
  <div>
    <div
      class="d-flex flex-column flex-grow-1"
      data-test="DesignSystem-InputMask--Wrapper"
    >
      <div
        class="Input Input--regular"
        data-test="DesignSystem-InputWrapper"
        role="presentation"
        style="min-width: 256px;"
      >
        <input
          autocomplete="off"
          class="Input-input Input-input--regular"
          data-test="DesignSystem-Input"
          placeholder="--:-- AM"
          type="text"
          value="08:44 PM"
        />
        <div
          class="Input-icon Input-iconWrapper--right"
        >
          <i
            class="material-icons material-icons-round Icon Input-icon--right"
            data-test="DesignSystem-Input--closeIcon"
            role="button"
            style="font-size: 16px; width: 16px;"
            tabindex="0"
          >
            close_round
          </i>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TimePicker component prop:validators snapshot 
  validators: false
 1`] = `
<body>
  <div>
    <div
      class="d-flex flex-column flex-grow-1"
      data-test="DesignSystem-InputMask--Wrapper"
    >
      <div
        class="Input Input--regular"
        data-test="DesignSystem-InputWrapper"
        role="presentation"
        style="min-width: 256px;"
      >
        <input
          autocomplete="off"
          class="Input-input Input-input--regular"
          data-test="DesignSystem-Input"
          placeholder="--:-- AM"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`TimePicker component prop:validators snapshot 
  validators: true
 1`] = `
<body>
  <div>
    <div
      class="d-flex flex-column flex-grow-1"
      data-test="DesignSystem-InputMask--Wrapper"
    >
      <div
        class="Input Input--regular"
        data-test="DesignSystem-InputWrapper"
        role="presentation"
        style="min-width: 256px;"
      >
        <input
          autocomplete="off"
          class="Input-input Input-input--regular"
          data-test="DesignSystem-Input"
          placeholder="--:-- AM"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</body>
`;
